app/Http/Controllers/Auth/OAuth2Controller.php

```diff
@@ -0,0 +1,203 @@
+            $session_driver = session()->get('link_oauth2_driver');
+            session()->forget('link_oauth2_driver');
+
+            // Check if authenticated
+            if (! $this->auth->guard()->check()) {
+                $errors = new MessageBag(['user' => [__('auth.failed')]]);
+
+                return redirect()->route('auth.login')
+                    ->withErrors($errors);
+            }
+
+            $user = $this->auth->guard()->user();
+
+            // Get the user info from OAuth2
+            try {
+                $socialite_user = Socialite::driver($session_driver)->user();
+            } catch (\Exception $exception) {
+                $this->alert->danger(trans('base.account.oauth2_link_failed'))->flash();
+
+                return redirect()->route('account');
+            }
+
+            $oauth2_socialite_id = $socialite_user->getId();
+
+            $new_ids = [];
+            $done = false;
+            // Replace if already exists
+            foreach (preg_split('~,~', $user->getAttributes()['oauth2_id']) as $id) {
+                if (Str::startsWith($id, $session_driver)) {
+                    $id = $session_driver . ':' . $oauth2_socialite_id;
+                    $done = true;
+                }
+                $new_ids = array_merge($new_ids, [$id]);
+            }
+
+            // Add if doesnt exist
+            if (! $done) {
+                $new_ids = array_merge($new_ids, [$session_driver . ':' . $oauth2_socialite_id]);
+            }
+
+            $oauth2_id = implode(',', $new_ids);
+
+            try {
+                $this->updateService->handle($user, compact('oauth2_id'));
+            } catch (\Exception $e) {
+                $this->alert->danger(trans('base.account.oauth2_link_failed'))->flash();
+
+                return redirect()->route('account');
+            }
+
+            $this->alert->success(trans('base.account.details_updated'))->flash();
+
+            return redirect()->route('account');
+        }
+
+        // Get the current OAuth2 user else redirect to auth page
+        if (! session()->has('oauth2_driver')) {
+            return redirect()->route('auth.login');
+        }
+
+        $session_driver = session()->get('oauth2_driver');
+        session()->forget('oauth2_driver');
+
+        try {
+            $socialite_user = Socialite::driver($session_driver)->user();
+        } catch (\Exception $exception) {
+            return redirect()->route('auth.login');
+        }
+
+        // The user's Id.
+        $oauth2_id = $socialite_user->getId();
+
+        try {
+            // Try to get the user
+            $user = $this->repository->findFirstWhere([['oauth2_id', 'LIKE', '%' . $session_driver . ':' . $oauth2_id . '%']]);
+
+            // Login
+            $this->auth->guard()->login($user);
+            if ($this->auth->guard()->check()) {
+                return redirect()->route('index');
+            }
+        } catch (\Exception $e) {
```

Comment:
There should never be a blind catch, and in the rare case it is necessary a comment needs to be included as to why it is there.